File: LeetCode/cpp/739_Daily_Temperatures.cpp
class Solution {
public:
	vector<int> dailyTemperatures(vector<int>& temperatures) {
		vector<int> answer(temperatures.size(), 0);
		stack<pair<int, int>> st;

		for (int i = 0; i < temperatures.size(); i++) {
			if (st.empty() || st.top().first > temperatures[i]) {
				st.push({ temperatures[i], i });
				continue;
			}

			while (!st.empty() && st.top().first < temperatures[i]) {
				pair<int, int> p = st.top();
				st.pop();

				answer[p.second] = i - p.second;
			}

			st.push({ temperatures[i], i });
		}

		while (!st.empty()) {
			pair<int, int> p = st.top();
			st.pop();

			answer[p.second] = 0;
		}

		return answer;
	}
};

File: LeetCode/java/739_Daily_Temperatures.java
import java.util.Stack;

class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] answer = new int[temperatures.length];
        Stack<int[]> stack = new Stack<>();

        for (int i = 0; i < temperatures.length; i++) {
            if (stack.isEmpty() || stack.peek()[0] > temperatures[i]) {
                stack.push(new int[]{temperatures[i], i});
                continue;
            }

            while (!stack.isEmpty() && stack.peek()[0] < temperatures[i]) {
                int[] p = stack.pop();

                answer[p[1]] = i - p[1];
            }

            stack.push(new int[]{temperatures[i], i});
        }

        while (!stack.isEmpty()) {
            int[] p = stack.pop();

            answer[p[1]] = 0;
        }

        return answer;
    }
}